### PR TITLE
feat: Add Option to Suppress Pre- and Post-Retrieve Events

### PR DESCRIPTION
**Pull request by klewis-sfdc**
### What does this PR do?
* Adds a property to ComponentSet RetrieveOptions to suppress the pre- and post- lifecycle events that are emitted when the ComponentSet's retrieve operation runs.

### What issues does this PR fix or reference?
* In the VSCE repo, work is underway to add source-tracking (STL) into the deploy* and retrieve* commands so that they can be used against scratch orgs ([#4597](https://github.com/forcedotcom/salesforcedx-vscode/pull/4597)).    

* In order to complete that work, two capabilities are needed:
  * VSCE needs to retrieve a file to a tmp location outside of the project
    * (Currently STL complains about the path used by os.tmpdir)
  * VSCE needs to be able to retrieve a file that has known conflicts after a deploy attempt that failed because of those conflicts (for line-by-line file diffing)

@W-12551198@

### Functionality Before
* No way to retrieve a ComponentSet without emitting pre- and post- retrieve events.
  * So STL is always being notified and reacting to the events 

### Functionality After
* Using the `suppressEvents` option, a retrieve call can be made without emitting the events, therefore bypassing STL behaviors

**Alternatives Considered:**
* Have STL not track files that are retrieved to a location outside of the project
  * Don't really want to make changes that try to tell STL what to track/not track
* Turn off listening for the events in STL temporarily
  * VSCE uses long running processes, would have to remember to re-enable listener, could get into trouble in cases of exceptions

### CODE REVIEW

File: test/collections/componentSet.test.ts [hunk 1]
```diff
@@ -208,6 +208,20 @@ describe('ComponentSet', () => {
...
+        await stubConnection();
+        await componentSet.retrieve({ output: '', usernameOrConnection: connection, suppressEvents: true });
+
+        let preAndPostRetrieveEventCount = 0;
+        lifecycleEmitStub.args.forEach((event) => {
+          if (event[0] === ('scopedPreRetrieve' || 'scopedPostRetrieve')) {
+            preAndPostRetrieveEventCount = preAndPostRetrieveEventCount + 1;
+          }
+        });
+        expect(preAndPostRetrieveEventCount).to.equal(0);
```
Comment (klewis-sfdc) [inline test/collections/componentSet.test.ts:223]: Only checking for the absence of Pre and Post retrieve events.

File: src/client/types.ts [hunk 2]
```diff
@@ -326,6 +326,10 @@ export interface RetrieveOptions {
    * Specifies whether to unzip the retrieved zip file. Only applies when `format: metadata`.
    */
   unzip?: boolean;
+  /**
+   * Specifies whether to suppress the <Pre|Post><Retrieve> events
+   */
+  suppressEvents?: boolean;
```
Comment (klewis-sfdc) [inline src/client/types.ts:332]: Only suppressing Pre and Post retrieve events.  Other events are still fired (like apiVersionRetrieve).

File: METADATA_SUPPORT.md [hunk 3]
```diff
@@ -557,6 +557,7 @@ v58 introduces the following new types.  Here's their current level of support
 |Ai4mSettings|✅||
 |CodeBuilderSettings|✅||
 |DataWeaveResource|✅||
+|DynamicFormsSettings|✅||
```
Comment (klewis-sfdc) [inline METADATA_SUPPORT.md:560]: Is this file auto-updated?

Comment (mshanemc) [reply METADATA_SUPPORT.md:560]: yes, from git hooks.

**Review by mshanemc — APPROVED** [review]

**Comment by mshanemc** [comment]
@klewis-sfdc looks good to me.  Can you have someone from VSCode QA and merge since y'all have a use case for it and we don't?

**Comment by klewis-sfdc** [comment]
Yep, we can do QA and merge.  Thanks, Shane.

**Review by gbockus-sf — APPROVED** [review]
thanks @mshanemc & @klewis-sfdc great stuff

**Pull request merged**